packages/vscode-dependencies-validation/src/diagnostics.ts

```diff
@@ -0,0 +1,42 @@
+import type { DiagnosticCollection, ExtensionContext } from "vscode";
+import { Range, Diagnostic, window, workspace, Uri } from "vscode";
+import { findDependencyIssues } from "@sap-devx/npm-dependencies-validation";
+import { set } from "lodash";
+import { NPM_DEPENDENCY_ISSUES_CODE } from "./constants";
+
+/**
+ * Analyzes the package.json text document for problems.
+ * @param doc package.json text document to analyze
+ * @param dependencyIssueDiagnostics diagnostic collection
+ */
+export async function refreshDiagnostics(
+  packageJsonPath: string,
+  dependencyIssueDiagnostics: DiagnosticCollection
+): Promise<void> {
+  const npmLsResult = await findDependencyIssues(packageJsonPath);
+
+  const diagnostics: Diagnostic[] = [];
+  if (npmLsResult.problems) {
+    diagnostics.push(
+      constructDiagnostic(npmLsResult.problems, packageJsonPath)
+    );
+  }
+
+  dependencyIssueDiagnostics.set(Uri.file(packageJsonPath), diagnostics);
+}
+
+function constructDiagnostic(
+  problems: string[],
+  packageJsonPath: string
+): Diagnostic {
+  const range = new Range(0, 0, 0, 10);
+  const diagnostic = new Diagnostic(
+    range,
+    problems.join("\n"),
+    0 // DiagnosticSeverity.Error
+  );
+  diagnostic.code = NPM_DEPENDENCY_ISSUES_CODE;
+  set(diagnostic, "packageJsonPath", packageJsonPath); // TODO: how to pass needed data to diagnostic ?
```

Comment:
on JS objects you can always add properties.
But it is uncertain you can rely on them staying there.
For example Diagnostics in LSP are passed serialized between server and client so 
additional properties would not be serialized.

Also if diagnostics are passed between the extension-host process and the main VSCode processes these additional properties may also be lost.

What is the use case that requires an additional property here?

Reply:
I need this packageJsonPath as input to command. The command is constructed based on a diagnostics.
I did not find a better way to pass it